File: languages/sample.ml
(* OCaml: a variant for order status, a record, and a fold over a list. *)
type status = Pending | Paid of float | Cancelled of string

type order = { number : int; total : float; status : status }

let describe = function
  | { number; status = Pending; _ } -> Printf.sprintf "#%d pending" number
  | { number; status = Paid on; _ } -> Printf.sprintf "#%d paid (%.0f)" number on
  | { number; status = Cancelled reason; _ } when reason <> "" -> Printf.sprintf "#%d cancelled: %s" number reason
  | { number; _ } -> Printf.sprintf "#%d cancelled" number

let revenue orders =
  List.fold_left (fun acc o -> match o.status with Paid _ -> acc +. o.total | _ -> acc) 0.0 orders

module StatusSet = Set.Make (String)

let () =
  let orders = [ { number = 1; total = 120.5; status = Paid 20260924. }
               ; { number = 2; total = 0.; status = Cancelled "duplicate" }
               ; { number = 3; total = 42.; status = Pending } ] in
  List.iter (fun o -> print_endline (describe o)) orders;
  Printf.printf "revenue: %.2f\n" (revenue orders)

File: languages/sample.fs
// F#: a discriminated union for orders and a pipeline over them.
module Sample

open System

type Status =
    | Pending
    | Paid of paidOn: DateTime
    | Cancelled of reason: string

type Order = { Number: int; Total: decimal; Status: Status }

let describe order =
    match order.Status with
    | Pending -> sprintf "#%d pending" order.Number
    | Paid on -> sprintf "#%d paid on %s" order.Number (on.ToString "yyyy-MM-dd")
    | Cancelled reason when reason.Length > 0 -> sprintf "#%d cancelled: %s" order.Number reason
    | Cancelled _ -> sprintf "#%d cancelled" order.Number

let orders =
    [ { Number = 1; Total = 120.50m; Status = Paid (DateTime(2026, 9, 24)) }
      { Number = 2; Total = 0m; Status = Cancelled "duplicate" }
      { Number = 3; Total = 42m; Status = Pending } ]

let revenue =
    orders
    |> List.filter (fun o -> match o.Status with Paid _ -> true | _ -> false)
    |> List.sumBy (fun o -> o.Total)

orders |> List.iter (describe >> printfn "%s")
printfn "revenue: %M" revenue
